### PR TITLE
bpo-46195: Do not add `Optional` in `get_type_hints`

### PR DESCRIPTION
**Pull request by sobolevn**
Useful related links:
- https://github.com/python/peps/pull/689
- https://github.com/python/typing/issues/275

https://bugs.python.org/issue46195

### CODE REVIEW

**Comment by merwok** [comment]
A note about commit messages / PR titles: `something does something` is ambiguous because it could be describing wrong behaviour that the code did before the change, or new correct behaviour.  Similar to the docstring convention (method columnize: `Display a list of blah by blah`), a good commit message or PR title should show what the _changes_ do:

- Fix extra space added by columnize (better than: columnize adds extra space (a bug))
- Add extra space in columnize (better than: columnize adds extra space (a feature))
- columnize now adds extra space when needed (same — present tense + `now` avoids ambiguity)

So for this PR (if I understand current title correctly) it could be:
- Do not add Optional in get_type_hints
- Make get_type_hints not add Optional

**Comment by sobolevn** [comment]
@merwok thanks, it is way more readable now! Wording (and writing in general) in English is something I really try to improve.

**Comment by sobolevn** [comment]
@Zac-HD do we rely on this behavior in `hypothesis`? 🤔

**Comment by Zac-HD** [comment]
> @Zac-HD do we rely on this behavior in `hypothesis`? 🤔

It's certainly observable by our users, and would probably make some of our tests fail. 

That said, I support the more consistent design proposed in the BPO issue, and we'll be fine carrying another clause in our type hints wrapper in Hypothesis. Probably important to reference the relevant PEP (section of 484, I think?) in the changelog though.

**Comment by pbryan** [comment]
Will this be applied in Python 3.11, or is it intended to be back-ported to previous versions?

**Comment by sobolevn** [comment]
In my opinion this should be 3.11+ only 🤔 
But, there might be other opinions of course.

**Comment by pbryan** [comment]
I have production code that currently assumes parameters that default to `None` are implicitly `Optional`; the required changes to accommodate this are trivial on my part, and I can certainly plan within the 3.11 time horizon.

File: Doc/library/typing.rst [hunk 1]
```diff
@@ -2030,9 +2030,7 @@ Introspection helpers
 
    This is often the same as ``obj.__annotations__``. In addition,
    forward references encoded as string literals are handled by evaluating
-   them in ``globals`` and ``locals`` namespaces. If necessary,
```
Comment (gpshead) [inline Doc/library/typing.rst:2033]: Add a versionchanged section mentioning the behavior change from previous versions.

**Review by gpshead — CHANGES_REQUESTED** [review]

File: Doc/library/typing.rst (hunk 1 above)
Comment (sobolevn) [reply Doc/library/typing.rst:2033]: Done! To make bot happy: I have made the requested changes; please review again

File: Misc/NEWS.d/next/Library/2021-12-30-21-38-51.bpo-46195.jFKGq_.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Now :func:`typing.get_type_hints` does not add ``Optional`` to types with
+explicit ``None`` default value.
+
+This was changed in https://github.com/python/peps/pull/689
```
Comment (Fidget-Spinner) [inline Misc/NEWS.d/next/Library/2021-12-30-21-38-51.bpo-46195.jFKGq_.rst]: Minor suggestion to improve clarity:
```suggestion
Now :func:`typing.get_type_hints` does not add ``Optional`` to types with
explicit ``None`` default value. This aligns to changes to PEP 484 in
https://github.com/python/peps/pull/689
```

**Review by Fidget-Spinner — COMMENTED** [review]
> In my opinion this should be 3.11+ only 🤔

Yeah. This is definitely a breaking change so it should be 3.11+ only. I wonder how many runtime checkers are affected or even use `get_type_hints` for that matter? @samuelcolvin can I trouble you for an opinion on this please? @Tinche too for cattrs too please. Thank you.

In short, `typing.get_type_hints` automatically wraps a type annotation with `Optional` if it sees a default `None`. This PR proposes to cease that behavior.

File: Misc/NEWS.d/next/Library/2021-12-30-21-38-51.bpo-46195.jFKGq_.rst (hunk 2 above)
Comment (Tinche) [reply Misc/NEWS.d/next/Library/2021-12-30-21-38-51.bpo-46195.jFKGq_.rst:4]: @Fidget-Spinner I'm totally fine with this, correct change in my opinion.

Just out of curiosity, does this change only apply to function parameters or class members too?

**Comment by Tinche** [comment]
> > In my opinion this should be 3.11+ only 🤔
> 
> Yeah. This is definitely a breaking change so it should be 3.11+ only. I wonder how many runtime checkers are affected or even use `get_type_hints` for that matter? @samuelcolvin can I trouble you for an opinion on this please? @Tinche too for cattrs too please. Thank you.
> 
> In short, `typing.get_type_hints` automatically wraps a type annotation with `Optional` if it sees a default `None`. This PR proposes to cease that behavior.

@Fidget-Spinner I'm totally fine with this, correct change in my opinion.

Just out of curiosity, does this change only apply to function parameters or class members too?

**Comment by sobolevn** [comment]
@Tinche for classes `Optional` was never added implictly (which is another point why we should fix this inconsistency).

Here are two examples, `main` and `3.9`:

```python                                                          
» ./python.exe
Python 3.11.0a3+ (heads/main:e13cdca0f5, Jan 11 2022, 12:43:49) [Clang 11.0.0 (clang-1100.0.33.16)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import typing
>>> class A:
...   x: int = None
... 
>>> typing.get_type_hints(A)
{'x': <class 'int'>}
```

```python
» python
Python 3.9.9 (main, Dec 21 2021, 11:35:28) 
[Clang 11.0.0 (clang-1100.0.33.16)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import typing
>>> class A:
...   x: int = None
... 
>>> typing.get_type_hints(A)
{'x': <class 'int'>}
```

**Comment by Tinche** [comment]
> @Tinche for classes `Optional` was never added implictly (which is another point why we should fix this inconsistency).
> 
> Here are two examples, `main` and `3.9`:
> 
> ```python
> » ./python.exe
> Python 3.11.0a3+ (heads/main:e13cdca0f5, Jan 11 2022, 12:43:49) [Clang 11.0.0 (clang-1100.0.33.16)] on darwin
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import typing
> >>> class A:
> ...   x: int = None
> ... 
> >>> typing.get_type_hints(A)
> {'x': <class 'int'>}
> ```
> 
> ```python
> » python
> Python 3.9.9 (main, Dec 21 2021, 11:35:28) 
> [Clang 11.0.0 (clang-1100.0.33.16)] on darwin
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import typing
> >>> class A:
> ...   x: int = None
> ... 
> >>> typing.get_type_hints(A)
> {'x': <class 'int'>}
> ```
> 
> However, I can't find a test case for this 🤔

Thanks for the info!

**Comment by sobolevn** [comment]
Thanks for the great question! 👍 

I've added this corner case to our tests in a separate PR: https://github.com/python/cpython/pull/30535

File: Doc/library/typing.rst [hunk 3]
```diff
@@ -2059,6 +2057,10 @@ Introspection helpers
    .. versionchanged:: 3.9
       Added ``include_extras`` parameter as part of :pep:`593`.
 
+   .. versionchanged:: 3.11
+      ``Optional`` annotation is not added implicitly anymore
+      when ``None`` default is used for a function argument.
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]: I like the wording in the removed docs better. Suggestion:

"Previously, ``Optional[t]`` was added for function and method annotations if a default value equal to ``None`` was set. Now the annotation is returned unchanged."

**Review by JelleZijlstra — COMMENTED** [review]
I like this change, just a comment on the docs.

File: Doc/library/typing.rst (hunk 3 above)
Comment (sobolevn) [reply Doc/library/typing.rst]: Great addition!

File: Doc/library/typing.rst (hunk 1 above)
Comment (AlexWaygood) [reply Doc/library/typing.rst:2033]: @sobolevn I think you need to post the message in a standalone message to make Bedevere happy, he's a pretty picky Knight :/

**Comment by samuelcolvin** [comment]
> > In my opinion this should be 3.11+ only thinking
> 
> Yeah. This is definitely a breaking change so it should be 3.11+ only. I wonder how many runtime checkers are affected or even use `get_type_hints` for that matter? @samuelcolvin can I trouble you for an opinion on this please? @Tinche too for cattrs too please. Thank you.
> 
> In short, `typing.get_type_hints` automatically wraps a type annotation with `Optional` if it sees a default `None`. This PR proposes to cease that behavior.

Thanks @Fidget-Spinner for asking my opinion, really useful to get a heads up about things like this. I don't *think* it will affect pydantic, if it does we can take care of it when adding support for 3.11.

I'm happy with this change provided it's no back-ported to other versions of python.

File: Misc/NEWS.d/next/Library/2021-12-30-21-38-51.bpo-46195.jFKGq_.rst [hunk 4]
```diff
@@ -0,0 +1,3 @@
+Now :func:`typing.get_type_hints` does not add ``Optional`` to types with
+explicit ``None`` default value. This aligns to changes to PEP 484 in
```
Comment (JelleZijlstra) [inline Misc/NEWS.d/next/Library/2021-12-30-21-38-51.bpo-46195.jFKGq_.rst]:
```suggestion
:func:`typing.get_type_hints` no longer adds ``Optional`` to parameters with
``None`` as a default. This aligns to changes to PEP 484 in
```

**Review by JelleZijlstra — APPROVED** [review]
@gvanrossum this is another one I'd like to merge.

@gpshead previously requested changes, but his request was addressed.

**Review by gvanrossum — APPROVED** [review]
IIUC, @gpshead mist still approve it.

**Review by gpshead — APPROVED** [review]

**Comment by Zac-HD** [comment]
Just re-found this issue because Hypothesis does indeed have a test for this 😂 

Let's mention this change in https://docs.python.org/3.11/whatsnew/3.11.html, not just the docs for `typing.get_type_hints()`.

**Pull request merged**